### PR TITLE
Add Solvent Extraction documentation

### PR DESCRIPTION
**Pull request by ArkoprabhoDasgupta**
## Changes proposed in this PR:
- Documentation is added to the model, property packages and the basic flowsheet
-

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.md file
   at the top level of this directory.

2. I represent I am authorized to make the contributions and grant the license. If my employer has
   rights to intellectual property that includes these contributions, I represent that I have
   received permission to make contributions and grant the required license on behalf of that
   employer.

### CODE REVIEW

File: src/prommis/solvent_extraction/solvent_extraction.py [hunk 1]
```diff
@@ -1,5 +1,87 @@
...
+
+Author: Arkoprabho Dasgupta
+
+The Solvent Extraction unit model is used to perform the solvent extraction unit operation.
+It represents a series of tanks, referred to as stages, through which the aqueous and organic
+phases are passed through, and then subsequent extraction of the desired components happen.
+
+This model is based on the MSContactor model, but it only accounts for the material balance.
+Any kind of energy balance, pressure balance and reaction kinetics is not considered in this 
+model. However since this model is based on MSContactor, the energy balance and pressure balances
```
Comment (andrewlee94) [inline src/prommis/solvent_extraction/solvent_extraction.py:14]: This is not correct. Whilst the default setting is False, the model does support energy and momentum balances for each stream (it does not support inter-stream heat transfer however).

File: src/prommis/solvent_extraction/solvent_extraction.py [hunk 2]
```diff
@@ -1,5 +1,87 @@
...
+
+The "property_package_args" configuration is any specific set of arguments that has to be passed to 
+the property block for the unit operation.
+
+The user can specify the direction of the flow of the stream through the stages through the 
+configuration "flow_direction". This is a Boolean type of configuration, with two inputs, based
+on the FlowDirection package. The default value is "FlowDirection.forward" which is the forward 
+direction, and the other is "FlowDirection.backward" which is the backward direction. 
+
+Since this model is based on MSContactor, the user has to give two additional arguments of
```
Comment (andrewlee94) [inline src/prommis/solvent_extraction/solvent_extraction.py:56]: As mentioned above, this is incorrect. The model does support these (even if you don't intend to use them right now), and the user can freely set these as they wish (assuming their property package supports the necessary properties).

File: src/prommis/solvent_extraction/sx_model_flowsheet_steady_state.py [hunk 3]
```diff
@@ -7,127 +15,141 @@
...
+m.fs.solex.mscontactor.organic_inlet_state[0].flow_vol.fix(62.01)
+
+"""
+Initialization of the model, which gives a good starting point.
+
+"""
+
+initializer = BlockTriangularizationInitializer(constraint_tolerance=1e-4)
+initializer.initialize(m.fs.solex)
+assert initializer.summary[m.fs.solex]["status"] == InitializationStatus.Ok
```
Comment (andrewlee94) [inline src/prommis/solvent_extraction/sx_model_flowsheet_steady_state.py:137]: It is generally not good practice to put `assert` statements in user facing code, as these will not give useful error messages if they fail.  `asserts` should generally be limited to tests or `try/except` statements,

**Review by andrewlee94 — CHANGES_REQUESTED** [review]
A couple of minor requests to change some of the docs. Whilst your current example does not have energy or momentum balances, the model as written does support them so the docs should reflect that.

Also, I see that you made some minor changes to the code but I don't see any changes to tests which implies that these liens of code are not tested. This can probably be left for another PR however.

File: src/prommis/solvent_extraction/solvent_extraction.py (hunk 1 above)
Comment (ArkoprabhoDasgupta) [reply src/prommis/solvent_extraction/solvent_extraction.py:14]: So I can remove the energy balance and pressure balance line, then. Will that be okay? I will keep the point that the configurations have to be set False in the stream config.

File: src/prommis/solvent_extraction/solvent_extraction.py (hunk 2 above)
Comment (ArkoprabhoDasgupta) [reply src/prommis/solvent_extraction/solvent_extraction.py:56]: Are you talking about lines 53-56?

**Comment by MarcusHolly** [comment]
You also need to add the solvent extraction directories to this file: https://github.com/prommis/prommis/blob/main/docs/api.rst in order for the documentation to appear.

File: src/prommis/solvent_extraction/sx_model_flowsheet_steady_state.py (hunk 3 above)
Comment (ArkoprabhoDasgupta) [reply src/prommis/solvent_extraction/sx_model_flowsheet_steady_state.py:137]: Okay I will remove it

**Comment by ArkoprabhoDasgupta** [comment]
> A couple of minor requests to change some of the docs. Whilst your current example does not have energy or momentum balances, the model as written does support them so the docs should reflect that.
> 
> Also, I see that you made some minor changes to the code but I don't see any changes to tests which implies that these liens of code are not tested. This can probably be left for another PR however.

@andrewlee94 can you explain the second point? Are you talking about the pytest file?

**Comment by ArkoprabhoDasgupta** [comment]
> You also need to add the solvent extraction directories to this file: https://github.com/prommis/prommis/blob/main/docs/api.rst in order for the documentation to appear.

I have added

File: src/prommis/solvent_extraction/solvent_extraction.py (hunk 1 above)
Comment (andrewlee94) [reply src/prommis/solvent_extraction/solvent_extraction.py]: You are confusing two separate things here. Your example property package does not include temperature and pressure, and thus for your examples you need to set these terms to False.

However, the SX model itself fully supports energy and momentum balances (as long as the property package has the necessary terms). Thus, this statement is incorrect. The model does support energy and momentum balances, and there is no need to set the arguments to False.

**Comment by andrewlee94** [comment]
@ArkoprabhoDasgupta Yes, I was referring to the pytest files. The fact that you can change these arguments without having to alos change the test files generally implies that things were not being fully tested before. Otherwise, a change in the user facing arguments should necessitate a change to the test files.

File: src/prommis/solvent_extraction/solvent_extraction.py (hunk 1 above)
Comment (ArkoprabhoDasgupta) [reply src/prommis/solvent_extraction/solvent_extraction.py]: Okay, I got your point. I will remove this term that you said

**Comment by ArkoprabhoDasgupta** [comment]
> @ArkoprabhoDasgupta Yes, I was referring to the pytest files. The fact that you can change these arguments without having to also change the test files generally implies that things were not being fully tested before. Otherwise, a change in the user facing arguments should necessitate a change to the test files.

@andrewlee94 I have changed the pytest file based on the latest solvent extraction flowsheet. Is there any other change you are talking about?

File: src/prommis/solvent_extraction/solvent_extraction.py (hunk 2 above)
Comment (ArkoprabhoDasgupta) [reply src/prommis/solvent_extraction/solvent_extraction.py:54]: I have removed this part also.

**Comment by andrewlee94** [comment]
@ArkoprabhoDasgupta I do not see any test files changed in this PR, so if you did make a change you have not pushed it. However, this is not necessary for this PR - I was commenting for future reference.

**Comment by ArkoprabhoDasgupta** [comment]
> @ArkoprabhoDasgupta I do not see any test files changed in this PR, so if you did make a change you have not pushed it. However, this is not necessary for this PR - I was commenting for future reference.

@andrewlee94 I checked the files changed in this PR and I understood what you are talking about. The pytest file is already updated in the main prommis repo. There was no additional changes required to be done in it, that is why you dont see it in this PR.

File: src/prommis/solvent_extraction/ree_og_distribution.py [hunk 4]
```diff
@@ -14,6 +22,20 @@
...
+    This is a property package for the organic phase solution of the solvent extraction
+    unit operation of the University of Kentucky pilot plant flowsheet.
+
+    This  includes the following components:
+
+    * Solvent: DEHPA
+    * Rare Earths: Sc, Y, La, Ce, Pr, Nd, Sm, Gd, Dy
+    * Impurities: Al, Ca, Fe
+
+    DEHPA is considered to be not involved in any kind of reaction.
```
Comment (agarciadiego) [inline src/prommis/solvent_extraction/ree_og_distribution.py:35]:
```suggestion
    DEHPA is not considered to be involved in any reaction.
```

File: src/prommis/solvent_extraction/solvent_extraction.py [hunk 5]
```diff
@@ -1,5 +1,77 @@
+"""
+Solvent Extraction Model
+
+========================
+
+Author: Arkoprabho Dasgupta
+
+The Solvent Extraction unit model is used to perform the solvent extraction unit operation.
+It represents a series of tanks, referred to as stages, through which the aqueous and organic
+phases are passed through, and then subsequent extraction of the desired components happen.
```
Comment (agarciadiego) [inline src/prommis/solvent_extraction/solvent_extraction.py:10]:
```suggestion
phases are passed, and the desired components are extracted subsequently.
```

File: src/prommis/solvent_extraction/solvent_extraction.py [hunk 6]
```diff
@@ -1,5 +1,77 @@
...
+-----------------------
+
+The user must specify the following configurations in a solvent extraction model to be able to 
+use it.
+
+The user must specify the aqueous feed input in the "aqueous_stream" configuration, with a 
+configuration defined with it which describes the properties of the aqueous feed.
+
+The user must specify the organic feed input in the "organic_stream" configuration, with a 
+configuration defined with it which describes the properties of the organic feed.
```
Comment (agarciadiego) [inline src/prommis/solvent_extraction/solvent_extraction.py:22]:
```suggestion
configuration that describes the organic feed's properties.
```

File: src/prommis/solvent_extraction/solvent_extraction.py [hunk 7]
```diff
@@ -1,5 +1,77 @@
...
+phases are passed through, and then subsequent extraction of the desired components happen.
+
+Configuration Arguments
+-----------------------
+
+The user must specify the following configurations in a solvent extraction model to be able to 
+use it.
+
+The user must specify the aqueous feed input in the "aqueous_stream" configuration, with a 
+configuration defined with it which describes the properties of the aqueous feed.
```
Comment (agarciadiego) [inline src/prommis/solvent_extraction/solvent_extraction.py:19]:
```suggestion
configuration that describes the properties of the aqueous feed.
```

File: src/prommis/solvent_extraction/solvent_extraction.py [hunk 8]
```diff
@@ -1,5 +1,77 @@
...
+The user must specify the organic feed input in the "organic_stream" configuration, with a 
+configuration defined with it which describes the properties of the organic feed.
+
+The number of stages in the solvent extraction process has to be specified by the user through 
+the "number_of_finite_elements" configuration. It takes an integer value.
+
+The material transfer can happen from either of the phases to the other. To specify the direction 
+of the transfer, the "aqueous_to_organic" configuration is to be used by the user. This is a boolean 
+configuration. The default value is True, which means the material transfer is happening from the 
+aqueous phase to the organic  phase, like in the loading operation. For scrubbing and stripping, the
```
Comment (agarciadiego) [inline src/prommis/solvent_extraction/solvent_extraction.py:30]:
```suggestion
aqueous phase to the organic phase, like in the loading operation. For scrubbing and stripping, the 
```

File: src/prommis/solvent_extraction/solvent_extraction.py [hunk 9]
```diff
@@ -1,5 +1,77 @@
...
+The material transfer can happen from either of the phases to the other. To specify the direction 
+of the transfer, the "aqueous_to_organic" configuration is to be used by the user. This is a boolean 
+configuration. The default value is True, which means the material transfer is happening from the 
+aqueous phase to the organic  phase, like in the loading operation. For scrubbing and stripping, the 
+reverse happens, so the value of the configuration will be False.
+
+Stream configurations
+---------------------
+
+Each of the feed streams has to have a dictionary which specifies the property packages and other
```
Comment (agarciadiego) [inline src/prommis/solvent_extraction/solvent_extraction.py:36]:
```suggestion
Each of the feed streams has to have a dictionary that specifies the property packages and other 
```

File: src/prommis/solvent_extraction/solvent_extraction.py [hunk 10]
```diff
@@ -1,5 +1,77 @@
...
+aqueous phase to the organic  phase, like in the loading operation. For scrubbing and stripping, the 
+reverse happens, so the value of the configuration will be False.
+
+Stream configurations
+---------------------
+
+Each of the feed streams has to have a dictionary which specifies the property packages and other 
+details as mentioned below.
+
+The "property_package" configuration is the property package which describes the state conditions
```
Comment (agarciadiego) [inline src/prommis/solvent_extraction/solvent_extraction.py:39]:
```suggestion
The "property_package" configuration is the property package that describes the state conditions 
```

File: src/prommis/solvent_extraction/ree_og_distribution.py (hunk 4 above)
Comment (ArkoprabhoDasgupta) [reply src/prommis/solvent_extraction/ree_og_distribution.py]: updated

File: src/prommis/solvent_extraction/solvent_extraction.py (hunk 5 above)
Comment (ArkoprabhoDasgupta) [reply src/prommis/solvent_extraction/solvent_extraction.py]: updated

File: src/prommis/solvent_extraction/solvent_extraction.py (hunk 6 above)
Comment (ArkoprabhoDasgupta) [reply src/prommis/solvent_extraction/solvent_extraction.py]: updated

File: src/prommis/solvent_extraction/solvent_extraction.py (hunk 7 above)
Comment (ArkoprabhoDasgupta) [reply src/prommis/solvent_extraction/solvent_extraction.py]: updated

File: src/prommis/solvent_extraction/solvent_extraction.py (hunk 8 above)
Comment (ArkoprabhoDasgupta) [reply src/prommis/solvent_extraction/solvent_extraction.py]: updated

File: src/prommis/solvent_extraction/solvent_extraction.py (hunk 9 above)
Comment (ArkoprabhoDasgupta) [reply src/prommis/solvent_extraction/solvent_extraction.py]: updated

File: src/prommis/solvent_extraction/solvent_extraction.py (hunk 10 above)
Comment (ArkoprabhoDasgupta) [reply src/prommis/solvent_extraction/solvent_extraction.py]: updated

**Review by MarcusHolly — APPROVED** [review]
LGTM

**Comment by lbianchi-lbl** [comment]
@ArkoprabhoDasgupta I've resolved a merge conflict in the `api.rst` file, and noticed that the head branch of this PR (i.e. the "source" branch) is your fork's `main`. I don't think this creates problems for this PR, but just as a reference for the future, we recommend that developers create a dedicated branch on their fork (i.e. in this case it could have been named `add-sx-docs`) instead of using their fork's `main`.

File: src/prommis/solvent_extraction/solvent_extraction.py [hunk 11]
```diff
@@ -1,5 +1,77 @@
...
+details as mentioned below.
+
+The ``property_package`` configuration is the property package that describes the state conditions 
+and properties of a particular stream.
+
+The ``property_package_args`` configuration is any specific set of arguments that has to be passed to 
+the property block for the unit operation.
+
+The user can specify the direction of the flow of the stream through the stages through the 
+configuration ``flow_direction``. This is a Boolean type of configuration, with two inputs, based
```
Comment (andrewlee94) [inline src/prommis/solvent_extraction/solvent_extraction.py:46]: This is configuration uses the ``FlowDirection`` ``Enum``, which has two possible values.

**Review by andrewlee94 — COMMENTED** [review]
One more minor request for a change.

**Review by andrewlee94 — APPROVED** [review]

**Pull request merged**